[코일] 수동 로또 미션 제출합니다.

takhyeongmin opened the pull request:
안녕하세요 리뷰어님! 

코일입니다..!

rebase가 제대로 안돼서 충돌 해결하느라고 시간이 지체됐네요!

궁금한 점이 있으면 DM해도 될까요 ?!

kimhyochan committed:
[docs] README 작성

kimhyochan committed:
[feat] 구매 개수 계산 메서드 작성

생성자로는 구매 금액으로부터 구매 개수를 산출하는
로직을 표현할 수가 없다고 판단하여 정적 팩토리 메서드 사용했습니다.

kimhyochan committed:
[feat] 구매 금액 입력

kimhyochan committed:
[feat] 구매 개수 출력

kimhyochan committed:
[refactor] 구매 개수 클래스에 getter 추가

kimhyochan committed:
[feat] 로또 넘버 생성 추가 및 예외 처리

로또 넘버의 유효성 검증을 한다.

kimhyochan committed:
[feat] 로또 넘버 리스트 일급 컬렉션 추가

로또 넘버들을 자료구조로 로또 넘버들을 관리하고 값 도출하는 용도

kimhyochan committed:
[feat] 중복없는 랜덤값 생성하는 클래스 작성

kimhyochan committed:
[refactor] 로또 숫자 enum 으로 변경

로또 숫자 재생성 비용을 enum 으로 변경함으로써 단 한번의 생성으로 객채 생성 비용 절약하였습니다.

kimhyochan committed:
[feat] 중복없는 무작위 숫자 6개를 생성하는 메소드 작성

kimhyochan committed:
[feat] 생성된 모든 LottoNumbers 출력

kimhyochan committed:
[feat] LottoNumbers의 목록을 가지고 있는 일급 컬렉션 생성

kimhyochan committed:
[feat] 당첨 번호와 보너스 볼을 관리하는 클래스 생성

int형의 번호를 알맞은 Enum 타입으로 파싱하는 메서드 생성

kimhyochan committed:
[feat] 당첨 번호 중복 예외 처리

kimhyochan committed:
[feat] 구매한 모든 로또 숫자들의 유효성을 검증하는 메서드 추가

kimhyochan committed:
[fix] WinningNumbers의 생성 중 중복 체크 중 추가되는 버그 수정

kimhyochan committed:
[feat] 지난 주 당첨번호와 보너스 볼을 입력받는 기능

kimhyochan committed:
[feat] 당첨 결과를 담는 enum 생성

kimhyochan committed:
[feat] 어플리케이션 메인 로직 생성

Commit:
Merge pull request #1 from hyochan-Kim/lotto

Lotto 풀리퀘 드립니다!

Commit:
[refact] test 폴더 루트 리팩토링

Commit:
[refact] Map 캐시 전략을 이용한 싱글톤 LottoNumber 생성

기존의 클래스를 한번만 생성하기 위해 사용했던
Enum LottoNumberGroup에서 LottoNumber 클래스로 변경하며
Map을 사용한 팩토리 메서드를 활용하기로 하였습니다.

같은 기능을 하지만 Enum 타입은 보일러 플레이트 코드가 너무
많아져 변경하게되었다.

Commit:
[refact] LottoNumber 타입 리팩토링

LottoNumber의 타입을 변경하면서 발생하는 에러를 리팩토링하였다.

Commit:
[refact] LottoNumbers, WinningNumbers 상속 리팩토링

LottoNumbers와 WinningNumbers의 중복 사항들을 LottoRound
클래스로 옮겨 중복을 제거하고 상속받게 리팩토링하였다.

Commit:
[refact] WinningNumbersTest 타입 리팩토링

LottoNumber의 타입을 변경하면서 발생하는 에러를 리팩토링하였다.

Commit:
[feat] GameResult 계산 기능

interface를 추출하여 알맞은 GameResult를 추출하는 기능을 추가하였다.

Commit:
[feat] 패키지 리팩토링

가독성을 위해 공통적인 로직을 가지고 있는 디렉토리 위치를 변경하였다.

Commit:
[feat] 테스트 패키지 리팩토링

가독성을 위해 공통적인 로직을 가지고 있는 디렉토리 위치를 변경하였다.

Commit:
[feat] GameResult 출력 기능 추가

GameResult를 출력하기 위해 GameResult 판다 논리를 담은
람다식 매개변수를 맞은 로또 넘버의 수로 바꿔
GameResult를 출력하기 용이하게 변경하였다.

Commit:
[feat] 수익률 계산

Commit:
[feat] 수익률 출력

Commit:
[feat] 당첨 번호와 로또 번호 중 맞는 수 카운트

WinnginNumbers에 당첨번호와 로또 번호 중 맞는 수를
카운팅하는 기능을 추가했다.

Commit:
[feat] 보너스 번호가 맞는지 확인하는 기능

Commit:
[fix] int 범위를 넘어가는 이익의 수익률 계산

int범위를 넘어가는 이익이 들어올 때 수익률 계산 에러 해결
double형으로 교체하였다.

Commit:
[refact] 랜덤넘버생성기 테스트 반복 추가

Commit:
[refact] main 메서드 분리 리팩토링

Commit:
[refact] inputView 발생 예외 커스텀

예외를 catch하여 알맞은 동작을 구동시키기 위해
커스텀 Exception을 생성하였다.

Commit:
[fix] 당첨 통계 금액 부동소수점 표기 수정

Commit:
[refact] AllLottoNnumbers에 GameResult 리스트 책임 부여

Application에서 가지고 있던 책임을 AllLottoNumbers에게 위임

Commit:
[fix] LottoNumber 정렬

Commit:
[refact] 자바 컨벤션

Commit:
[refact] 테스트 코드 리팩토링

Commit:
[refact] 테스트 코드 리팩토링

Commit:
[refact] LottoRound 자료형 접근지정자 변경

Commit:
[refact] AllLottoNumbers Null 체크 로직 변경

Commit:
[feat] 구매 금액 이하의 값 입력 시 예외 처리

Commit:
[refact] 천 단위 구별을 위한 "_" 추가

Commit:
[refact] Yield, PurchaseNumber를 Money로 리팩토링

수익률을 의미하는 Yield, 로또 라운드 수를 의미하는 PurchaseNumber가
Money라는 이름의 금액으로 추상화될 수 있다고 판단하여
하나의 클래스에서 책임지도록 변경하였다.

Commit:
[refact] AllLottoNumbers, LottoNumbers, LottoNumber 클래스명 변경

구매한 모든 로또 라운드의 수를 가지고 있는 AllLottoNumbers와
라운드 하나의 로또 숫자를 가지고 있는 LottoNumbers와
1부터 45의 로또를 구성하는 숫자를 가지고 있는 LottoNumber의
작명이 혼동을 일으킨다고 판단하여 각 LottoRounds, LottoRound로
변경하였다.

또한 LottoRound를 LottoRoundFormat으로 변경하여 한 라운드의
규격을 나타내려는 의도를 보이도록 하였다.

Commit:
[refact] 로또 번호 생성 클래스 추출

LottoRoundsGenerator 클래스를 통해 로또 번호 생성하는 클래스를
추출하여 책임 부여하였다.

Commit:
[refact] LottoRound Set 자료구조 변경

중복을 허용하지 않는 점에서 List 보다 Set이 중복에 대한
예외를 처리하기 용이하다고 판단하여 Set으로 변경하였다.

Commit:
[refact] boolean 메서드 네이밍 변경

보너스 번호를 가지고 있는지 확인하는 메서드에서 is에서
가지고 있다를 의미하는 has로 메서드 명을 변경하였습니다.

Commit:
[refact] 게임 결과 도출 테스트 추가

Commit:
[refact] LottoGame 클래스 추가

Main은 입출력과 게임 실행을 이어주는 기능만을 담당해야한다고
생각하여 게임의 흐름을 제어하는 LottoGame을 생성하였다.

Commit:
[refact] LottoRoundFormat 삭제

WinnningNumbers, LottoRound 에서 동일하게 사용하는 여섯개의
숫자를 나타내는 자료형을 둘다 다루는 것에서 LottoRoundFormat을
상속받는 형태로 구현하였으나,

상속보다는 유연한 객체 조립으로 해당 기능을 리팩토링하였다.

Commit:
[refact] 컨벤션 리팩토링

Commit:
[docs] README.md 작성

Commit:
[docs] README.md 기능 변경 추가

Commit:
[feat] 수동 구매 로또 수량 입력

Commit:
[feat] 수동 구매 로또를 구성하는 수 입력

Commit:
[feat] 수동 로또 구매 수량이 전체 금액 이상인 경우 예외 처리

Commit:
[feat] 수동 로또 입력 기능

LottoRoundsGeneratator에서 수동 로또로 입력 받은 부분을 제외하고
자동 로또를 생성하는 로직으로 변경

LottoGame의 initialize() 가 실행될 때, 수동으로 입력된 로또가
포함되도록 매개변수로 수동 로또를 입력받게 변경

Commit:
[refact] 수동 로또 입력 시 헤더 수정 및 메서드 추출

당첨 번호를 입력받는 로직과 수동 로또를 입력받는 로직에
중복이 있어 메서드 추출을 통해 중복을 제거하였다.

Commit:
[feat] 구매 로또 수 출력

Commit:
[refact] 당첨통계 출력문 개선

기존의 StringBuilder를 통합 문자열 결합이 복잡하다고 판단해 변경

Commit:
[refact] Money / GameResults 수익률 계산 메서드 의존 관계 변경

Money에 작성된 수익률 계산 메서드가 Money보다 GameResults에서
더많은 정보를 사용하고 있다고 판단하여 GameResults에서 Money를
매개변수로 전달받도록 변경

Commit:
[refact] 컨벤션 리팩토링

Commit:
merge

kingbbode commented on src/main/java/lotto/domain/result/Money.java:
```diff
@@ -5,31 +5,34 @@
 import java.util.Objects;
 
 public class Money {
```
`Money` 가 로또의 가격을 알고 있는 것이 맞을까요?
`Money` 는 말 그대로 소유하고 있는 돈을 관리하는 기능에만 집중하고, 그 정보를 제공해주는 것은 어떨까요?
`Money` 는 `Lotto` 에 의존하지 않는 투명한 객체 역할을 충분히 수행할 수 있을 것 입니다 :)

그렇다면 로또 가격을 알고 있고, 로또를 발행하며 소유하고 있는 돈을 차감해줄 역할을 가진 객체가 필요하겠네요.

kingbbode commented on src/main/java/lotto/domain/LottoGame.java:
```diff
@@ -19,13 +19,13 @@ private LottoGame(Money money, LottoRounds lottoRounds) {
 
-    public static LottoGame initialize(Money money) {
-        LottoRounds lottoRounds = createAllLottoNumbers(money);
+    public static LottoGame initialize(Money money, List<LottoRound> manualLottos) {
```
생성시점에 `manualLottos` 를 받는 것이 옳은 것일까요?
`LottoGame` 의 역할은 무엇이고, 언제 어떤 정보를 취급해야할지 다시 한번 고민해보는 것은 어떨까요?

크게 초기화, 티켓 발행, 결과 생성의 3가지 흐름을 제공할 수 있지 않을까합니다!

kingbbode reviewed (changes requested):
안녕하세요 :) 
아쉬운 점이 남아 코멘트 작성하였습니다.
확인 부탁드려요!

Commit:
[docs] 피드백 개선 기능 목록 추가

Commit:
[feat] Money에 로또 티켓값을 제거하는 함수 추가

Commit:
[refact] Money 생성자 변경

생성자 매개변수에서 최저값을 지정해주는 형태로 변경,
생성자에서 유효성을 검증하고, Money는 TICKET_PRICE와 같은 값을
모르고서도 온전히 돈 관리에 대한 책임을 지도록 변경

Commit:
[refact] 수동 로또 금액 유효성 검증

Lotto Ticket값 없이도 수동 로또 금액을 검증할 수 있도록 매개변수
변경

Commit:
[feat] Money 차감 가능 확인 기능 추가

Commit:
[feat] LottoRounds들을 합치는 기능 추가

LottoRounds들을 합쳐서 새로운 LottoRounds를 만드는 기능 추가

Commit:
[feat] 수동, 자동 로또를 발행하는 Generator 생성

로또를 생성하는 Generator의 책임을 분할하여 수동, 자동 로또를
생성하는 클래스 생성

Commit:
[feat] LottoGame 메서드 static화

Commit:
[refact] LottoRoundsGenerator Test 세분화

Commit:
[refact] LottoRounds combine 함수 이름 재정의

Commit:
[refact] 수익률 계산 불필요 매개변수 제거

Commit:
[refact] 컨벤션 리팩토링

kingbbode reviewed (approved):
리뷰가 많이 늦어져서 죄송합니다.
이만 머지하겠습니다!
수고하셨습니다 :)

kingbbode commented:
개인적인 사정으로 리뷰가 많이 늦어졌습니다. 이번에 다 못한 리뷰는 다음에 다른 방법으로 최대한 도움이 되실 수 있도록 하겠습니다! 리뷰가 늦어 죄송합니다.
피드백드리고 제대로 리뷰를 못드렸으니, 코로나가 잠잠해지면 바로옆 건물에 있으니 찾아가서라도 리뷰를 해드리겠습니다 ㅜㅜ
